addon/src/webextension/lib/browserAction.js

```diff
@@ -61,11 +61,34 @@ async function updateBadgeTextOnNew(topic, { experiments, news_updates }) {
     return dt >= clicked;
   });
 
-  const newsUpdates = (news_updates || []).filter(update => {
-    const dt = new Date(update.published).getTime();
-    return dt >= clicked;
+  // check for port number on local, we need to strip it off
+  // to properly fetch cookies.
+  const portIndex = getCurrentEnv().baseUrl.indexOf(':8000');
+  const baseUrl = (portIndex > -1) ?
+        getCurrentEnv().baseUrl.substring(0, portIndex)
+        : getCurrentEnv().baseUrl.indexOf(':8000');
+
+  let lastViewed = 0;
+  const cookie = await browser.cookies.get({
+    url: baseUrl,
+    name: 'updates-last-viewed-date'
   });
 
+  if (cookie) lastViewed = cookie.value;
+
+  /* only show badge for news update if:
+   * - has the "major" key
+   * - update has been published in the past two weeks
+   * - update has not been "seen" by the frontend (lastViewed)
+   * - update has not been "seen" by the addon (clicked)
+   */
+  const TWO_WEEKS = 2 * 7 * 24 * 60 * 60 * 1000;
+  const twoWeeksAgo = Date.now() - TWO_WEEKS;
+  const newsUpdates = (news_updates || []).filter((u) => u.major)
+        .filter((u) => new Date(u.published).getTime() >= twoWeeksAgo)
```

Comment:
They are different dates in each instance, this is in a `.filter` call

Reply:
between these two I think the existing version is more readable 
```js
const newsUpdates = (news_updates || []).filter((u) => u.major)                                                                                                                                               
         .filter((u) => new Date(u.published).getTime() >= twoWeeksAgo)                                                                                                                                          
         .filter((u) => new Date(u.published).getTime() >= new Date(lastViewed))                                                                                                                                 
        .filter((u) => new Date(u.published).getTime() >= clicked);                                                                                                                                             
 ```
```js                                                                                                                                                                                                                
   const newsUpdates = (news_updates || []).map(u => {                                                                                                                                                           
     const pubDate = new Date(u.published).getTime();                                                                                                                                                            
     if (u.major &&                                                                                                                                                                                              
         pubDate >= twoWeeksAgo &&                                                                                                                                                                               
         pubDate >= new Date(lastViewed) &&                                                                                                                                                                      
         pubDate >= clicked) return u;                                                                                                                                                                           
 });
```

Reply:
You're literally calling that code 3 times for every element in the list... I don't think it is premature optimization. If you put const pubDate above const newsUpdates, you will only call it once instead of 3 * news_updates.length.

Reply:
Never mind, I thought published only had to be parsed once, but it has to be parsed for each element of news_updates. Carry on.